sirepo/runner_daemon/docker_process.py

```diff
@@ -0,0 +1,161 @@
+# -*- coding: utf-8 -*-
+u"""sirepo package
+
+:copyright: Copyright (c) 2018 RadiaSoft LLC.  All Rights Reserved.
+:license: http://www.apache.org/licenses/LICENSE-2.0.html
+"""
+from __future__ import absolute_import, division, print_function
+
+from pykern import pkcollections
+from pykern import pkio
+from pykern.pkdebug import pkdp, pkdc, pkdlog, pkdexc
+from sirepo.template import template_common
+import contextlib
+import docker
+import functools
+import re
+import requests
+import shlex
+import shutil
+import trio
+
+# How often threaded blocking operations should wake up and check for Trio
+# cancellation
+_CANCEL_POLL_INTERVAL = 1
+
+# Global connection pools (don't worry, it won't actually make any connections
+# until we start using it)
+_DOCKER = docker.from_env()
+
+def _container_name(run_dir, job_type):
+    # Docker container names have to start with [a-zA-Z0-9], and after that
+    # can also contain [_.-].
+    return 'sirepo-{}-{}'.format(job_type, str(run_dir).replace('/', '-'))
+
+
+# Helper to call container.wait in a async/cancellation-friendly way
+async def _container_wait(container):
+    while True:
+        try:
+            return await trio.run_sync_in_worker_thread(
+                functools.partial(container.wait, timeout=_CANCEL_POLL_INTERVAL),
+            )
+        # ReadTimeout is what the documentation says this raises.
+        # ConnectionError is what it actually raises.
+        # We'll catch both just to be safe.
+        except (requests.exceptions.ReadTimeout, requests.exceptions.ConnectionError):
+            pass
+
+
+async def _clear_container(name):
+    try:
+        old_container = await trio.run_sync_in_worker_thread(
+            _DOCKER.containers.get, name,
+        )
+    except docker.errors.NotFound:
+        pass
+    else:
+        pkdlog('found stale container {}; removing', name)
+        await trio.run_sync_in_worker_thread(old_container.remove)
```

Comment:
Not for any good reason :-). And if we genuinely fail to remove a container with the name we want to use, then the next step is going to fail regardless. But looking at it again, I guess it would actually make the code a bit simpler to allow `NotFound` from both `containers.get` or `container.remove`, and that does remove a theoretical race condition that we'd never hit in practice.